### PR TITLE
Major library refactoring

### PR DESCRIPTION
**Pull request by MarcAntoine-Arnaud**

### CODE REVIEW

File: lib/remote_docker.ex [hunk 1]
```diff
@@ -0,0 +1,16 @@
+defmodule RemoteDockers do
+  @moduledoc """
+  Elixir wrapper to drive Docker nodes.
+
+  Before try this library, be sure your docker machine will serve the API on HTTP (on some OS, default is on a socket).
```
Comment (valnoel) [inline lib/remote_docker.ex]: _Before using this library, ensure your Docker instance is serving the HTTP API (by default, the service is generally listening to a socket)_

File: lib/remote_docker.ex [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+  This library support the connection with SSL certificat.
```
Comment (valnoel) [inline lib/remote_docker.ex]: _This library support __s__ the connection with SSL certificat (HTTPS)._

File: lib/host_config.ex [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  default:
+  ```
+  hostname: "localhost"
+  port: 2376
+  ```
+  """
+  def build() do
+    %RemoteDockers.HostConfig{
+      hostname: "localhost",
+      port: 2376
```
Comment (valnoel) [inline lib/host_config.ex]: Couldn't we put this default port value as a \@constant ?

File: lib/container.ex [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  """
+  def get_status!(%RemoteDockers.Container{} = container) do
+    response =
+      Client.build_endpoint(@containers_uri, container.container_id <> "/json")
+      |> Client.build_uri(container.host_config)
+      |> Client.get!()
+
+    case response.status_code do
+      200 -> response.body["State"]["Status"]
+      _ -> raise "unable to retrieve container"
```
Comment (valnoel) [inline lib/container.ex:156]: _"unable to retrieve container __status__"_

File: lib/container.ex [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+  @enforce_keys [:container_id, :host_config]
+  defstruct [:container_id, :host_config]
+
+  @containers_uri "/containers"
+
+  @doc """
+  List running containers
+  """
+  def list!(host_config) do
```
Comment (valnoel) [inline lib/container.ex]: Shouldn't we force these functions arguments types?

File: lib/host_config.ex (hunk 3 above)
Comment (MarcAntoine-Arnaud) [reply lib/host_config.ex]: don't change anything, it's used only here.
I use @constant if it's used many times.

File: lib/container.ex (hunk 4 above)
Comment (MarcAntoine-Arnaud) [reply lib/container.ex:156]: may depends on return code.
-> 404 means no container found ...
as a quick development I put this message for that.

File: lib/container.ex (hunk 5 above)
Comment (MarcAntoine-Arnaud) [reply lib/container.ex]: We should test typespec to improve that.

File: test/container_test.exs [hunk 6]
```diff
@@ -3,15 +3,19 @@ defmodule RemoteDockers.ContainerTest do
   alias RemoteDockers.{Container, HostConfig}
   doctest RemoteDockers.Container
 
+  @host_config HostConfig.build(
+    "https://192.168.99.100",
+    "/path/to/cert.pem",
+    "/path/to/key.pem"
+  )
```
Comment (MarcAntoine-Arnaud) [inline test/container_test.exs]: unit tests can never works with that.
We need to load configuration from config/test.exs

File: test/image_test.exs [hunk 7]
```diff
@@ -3,9 +3,14 @@ defmodule RemoteDockers.ImageTest do
   alias RemoteDockers.{Image, HostConfig}
   doctest RemoteDockers.Image
 
-  test "list containers" do
-    host_config = HostConfig.build("https://192.168.99.100", "/Users/marco/.docker/machine/certs/cert.pem", "/Users/marco/.docker/machine/certs/key.pem")
-    images = Image.list!(host_config)
+  @host_config HostConfig.build(
+    "https://192.168.99.100",
+    "/path/to/cert.pem",
+    "/path/to/key.pem"
```
Comment (MarcAntoine-Arnaud) [inline test/image_test.exs]: idem, load from test config

File: lib/image.ex [hunk 8]
```diff
@@ -69,4 +94,21 @@ defmodule RemoteDockers.Image do
...
   end
+
+  defp to_pull_status(steps) do
+    steps
+    |> Enum.map_reduce("", fn step, acc ->
+        acc
+          |> get_id(step["id"])
+          |> put_id(step)
+      end)
+    |> elem(0)
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]:
```elixir
|> List.first
```

File: lib/image.ex [hunk 9]
```diff
@@ -69,4 +94,21 @@ defmodule RemoteDockers.Image do
       virtual_size: image["VirtualSize"]
     }
   end
+
+  defp to_pull_status(steps) do
+    steps
+    |> Enum.map_reduce("", fn step, acc ->
+        acc
+          |> get_id(step["id"])
+          |> put_id(step)
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]: don't indent, align with acc

File: lib/image.ex [hunk 10]
```diff
@@ -69,4 +94,21 @@ defmodule RemoteDockers.Image do
...
+      end)
+    |> elem(0)
+  end
+
+  defp get_id(acc, id) when is_nil(id), do: acc
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]:
```elixir
defp get_id(acc, nil) do: acc
```

File: lib/image.ex [hunk 11]
```diff
@@ -69,4 +94,21 @@ defmodule RemoteDockers.Image do
       virtual_size: image["VirtualSize"]
     }
   end
+
+  defp to_pull_status(steps) do
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]: what's the goal of this method ?

Comment (MarcAntoine-Arnaud) [reply lib/image.ex]: `format_status` ?

File: lib/image.ex [hunk 12]
```diff
@@ -54,6 +54,31 @@ defmodule RemoteDockers.Image do
...
+      HostConfig.get_options(host_config)
+
+    response =
+      Client.build_endpoint(@images_uri, "/create?fromImage=" <> name)
+      |> Client.build_uri(host_config)
+      |> Client.post!([], options)
+
+    case response.status_code do
+      200 -> to_pull_status(response.body)
+      _ -> raise "unable to pull a " <> name <> " image"
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]:
```
"unable to pull image with name: " <> name
```

File: lib/client.ex [hunk 13]
```diff
@@ -17,8 +17,14 @@ defmodule RemoteDockers.Client do
...
-    body
-    |> Poison.decode!
+    if String.contains?(body, "}\r\n{") do
+      body
+      |> String.split(~r/\r\n/, trim: true)
+      |> Enum.map(fn(part) -> Poison.decode!(part) end)
+    else
+      body
+      |> Poison.decode!
+    end
```
Comment (MarcAntoine-Arnaud) [inline lib/client.ex]: test with something like:
```elixir
body
|> String.split("\n")
|> Stream.map(&String.trim/1)
|> Stream.with_index
|> Stream.map(fn ({line, _index}) -> Poison.decode!(line) end)
|> Stream.run
```

File: lib/image.ex [hunk 14]
```diff
@@ -3,15 +3,52 @@ defmodule RemoteDockers.Image do
...
-    Client.build_endpoint(@images_uri)
-    |> Client.build_uri(host_config)
-    |> Client.get!([], HostConfig.get_options(host_config))
+    response =
+      Client.build_endpoint(@images_uri)
+      |> Client.build_uri(host_config)
+      |> Client.get!([], HostConfig.get_options(host_config))
+
+    case response.status_code do
+      200 -> Enum.map(response.body, fn(image) -> to_image(image, host_config) end)
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]:
```elixir
200 ->
  Enum.map(response.body, fn(image) ->
    to_image(image, host_config)
  end)
```

File: lib/image.ex [hunk 15]
```diff
@@ -3,15 +3,52 @@ defmodule RemoteDockers.Image do
...
+      _ -> raise "unable to list images"
+    end
+  end
+
+
+  defp to_image(%{} = image, %HostConfig{} = host_config) do
```
Comment (MarcAntoine-Arnaud) [inline lib/image.ex]: %{} is not required. We will define that with typespec.

**Pull request merged**